### PR TITLE
Put some @test's into the test for #19963

### PR DESCRIPTION
**Pull request by thofma**
Address https://github.com/JuliaLang/julia/pull/19969#pullrequestreview-16384757.

### CODE REVIEW

File: test/core.jl [hunk 1]
```diff
@@ -4866,8 +4866,8 @@ end
 # issue #15240
 
 p15240 = ccall(:jl_realloc, Ptr{Void}, (Ptr{Void}, Csize_t), C_NULL, 10)
-ccall(:jl_free, Void, (Ptr{Void}, ), p15240)
+@test ccall(:jl_free, Void, (Ptr{Void}, ), p15240) == nothing
```
Comment (tkelman) [inline test/core.jl]: `=== nothing` will also check the type

**Comment by StefanKarpinski** [comment]
Why the 👎  @yuyichao? I explicitly asked for this change.

**Comment by yuyichao** [comment]
While I agree making test count more accurate is good, I don't think we should be putting in these workarounds.

1. If we want this in general, we should do https://github.com/JuliaLang/julia/issues/18780 instead.

2. As a workaround without https://github.com/JuliaLang/julia/issues/18780, this is not general at all.

    `@test isa(ex, Any)` would be a more general solution but it's ugly enough that I think anyone who want to fix test count should just implement the more general solution instead.

3. Using these workaround achieves exactly the opposite goal of https://github.com/JuliaLang/julia/issues/18780.

    Other than a test count, which I think is less important, https://github.com/JuliaLang/julia/issues/18780 makes it clear what the line is testing. It is much less clear that `@test ex === nothing` or `@test is(ex, Any)` actually want to test `ex` doesn't crash or throw an error.

4. The first `nothing` check isn't even on the line to test. The line to be tested is the previous line.

**Comment by tkelman** [comment]
we have `@test_nowarn` now

**Comment by yuyichao** [comment]
A little overkill and a little strange but yeah, I that's a much better one to use than a dummy test on the result.

**Comment by thofma** [comment]
I will change it to use `@test_nowarn`.

By the way, is the following behavior intended? I did not expect it after reading the documentation for `@test_nowarn`.
```julia
julia> @test_nowarn p = 2 + 2
4

julia> @test p == 4
Error During Test
  Test threw an exception of type UndefVarError
  Expression: p == 4
  UndefVarError: p not defined
```
(I know I can work around it using `@test_nowarn eval(Main, :(begin p = 2 + 2 end))`.)

**Comment by yuyichao** [comment]
`@test_*` introduces new scope.

Don't use `eval` for this and especially don't use `eval` in `Main` in the test unless absolutely necessary.
You can just put the `@test_nowarn` around a `begin` block with multiple statements.

**Comment by thofma** [comment]
Yes, I know I can use it with `begin` blocks and multiple statements. But if there are multiple statments and it fails, I don't know which statement failed.

**Comment by yuyichao** [comment]
In that case since it's just introducing a new scope you can do whatever necessary to set the variable in the right scope. A `@testset` should work, for example.

```jl
julia> @testset "aaa" begin
           local a
           @test (a = 1) == 1
           @test a == 1
       end
Base.Test.DefaultTestSet("aaa",Any[Test Passed
,Test Passed
],false)
```

**Comment by thofma** [comment]
Probably I did not understand it correctly, but it does not work with `@test_nowarn`.
```julia
julia> Base.Test.@testset "aaa" begin
                  local a
                  Base.Test.@test_nowarn (a = 1) == 1
                  Base.Test.@test_nowarn a == 1
              end
aaa: Error During Test
  Got an exception of type UndefVarError outside of a @test
  UndefVarError: a not defined
```

**Comment by yuyichao** [comment]
.... This is a bug in `@test_nowarn` ....... It's missing an `esc`

i.e. it should be defined as

```
macro test_nowarn(expr)
    quote
        @test_warn r"^(?!.)"s $(esc(expr))
    end
end
```

**Comment by thofma** [comment]
Ok. I just go for the working solution

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
PR for `@test_warn` bug: https://github.com/JuliaLang/julia/pull/20017.

**Pull request merged**